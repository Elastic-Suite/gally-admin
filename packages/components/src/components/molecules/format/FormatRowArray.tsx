import React from 'react'
import { IMultipleValueFormat } from '@elastic-suite/gally-admin-shared'
import FormatArray from '../../atoms/format/FormatArray'
import FormatText from '../../atoms/format/FormatText'

interface IProps {
  values: string[]
  multipleValueFormat?: IMultipleValueFormat
}

function FormatRowArray(props: IProps): JSX.Element {
  const { values, multipleValueFormat } = props
  const content: string | string[] = multipleValueFormat?.separator
    ? values.join(multipleValueFormat.separator)
    : values

  if (!multipleValueFormat?.maxCount) {
    return (
      <div>
        {content instanceof Array
          ? content.map((item) => <div key={item}>{item}</div>)
          : content}
      </div>
    )
  }

  let maxChar = values.reduce(
    (acc, item, index) =>
      index < multipleValueFormat.maxCount ? acc + item.length : acc,
    0
  )
  maxChar =
    maxChar +
    (multipleValueFormat.maxCount - 1) * multipleValueFormat?.separator?.length

  return (
    <div>
      {content instanceof Array ? (
        <FormatArray values={content} max={multipleValueFormat.maxCount} />
      ) : (
        <FormatText
          name={content}
          max={maxChar}
          toolTip
          firstLetterUpp={false}
        />
      )}
    </div>
  )
}

export default FormatRowArray
