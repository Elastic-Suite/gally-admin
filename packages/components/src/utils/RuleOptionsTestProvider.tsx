import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
