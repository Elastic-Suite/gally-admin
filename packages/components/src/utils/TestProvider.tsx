import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
