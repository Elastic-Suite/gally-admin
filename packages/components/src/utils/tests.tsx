import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
