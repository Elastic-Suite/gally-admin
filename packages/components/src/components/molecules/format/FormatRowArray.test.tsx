import React from 'react'

import { renderWithProviders } from '../../../utils/tests'

import FormatRowArray from './FormatRowArray'

describe('FormatRowArray match snapshot', () => {
  it('FormatRowArray Full', () => {
    const { container } = renderWithProviders(
      <FormatRowArray
        values={['blazer', 'jacket', 'trousers', 'pants']}
        multipleValueFormat={{ separator: ', ', maxCount: 2 }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('FormatRowArray maxCount=1', () => {
    const { container } = renderWithProviders(
      <FormatRowArray
        values={['blazer', 'jacket', 'trousers', 'pants']}
        multipleValueFormat={{ separator: ', ', maxCount: 1 }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('FormatRowArray without maxCount', () => {
    const { container } = renderWithProviders(
      <FormatRowArray
        values={['blazer', 'jacket', 'trousers', 'pants']}
        multipleValueFormat={{ separator: ', ', maxCount: 1 }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('FormatRowArray without separator', () => {
    const { container } = renderWithProviders(
      <FormatRowArray
        values={['blazer', 'jacket', 'trousers', 'pants']}
        multipleValueFormat={{ maxCount: 1 }}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
