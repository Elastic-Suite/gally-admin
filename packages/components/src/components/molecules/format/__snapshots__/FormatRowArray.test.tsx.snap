// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatRowArray match snapshot FormatRowArray Full 1`] = `
<div>
  <div>
    <span
      aria-label="blazer, jacket, trousers, pants"
      class=""
      data-mui-internal-clone-element="true"
      style="display: inline-block;"
    >
      blazer, jacket...
    </span>
  </div>
</div>
`;

exports[`FormatRowArray match snapshot FormatRowArray maxCount=1 1`] = `
<div>
  <div>
    <span
      aria-label="blazer, jacket, trousers, pants"
      class=""
      data-mui-internal-clone-element="true"
      style="display: inline-block;"
    >
      blazer...
    </span>
  </div>
</div>
`;

exports[`FormatRowArray match snapshot FormatRowArray without maxCount 1`] = `
<div>
  <div>
    <span
      aria-label="blazer, jacket, trousers, pants"
      class=""
      data-mui-internal-clone-element="true"
      style="display: inline-block;"
    >
      blazer...
    </span>
  </div>
</div>
`;

exports[`FormatRowArray match snapshot FormatRowArray without separator 1`] = `
<div>
  <div>
    <div>
      <span
        class=""
        data-mui-internal-clone-element="true"
        style="display: inline-block;"
      >
        blazer
        ...
      </span>
    </div>
  </div>
</div>
`;
